service/src/main/java/bio/terra/cbas/runsets/results/RunCompletionHandler.java

```diff
@@ -0,0 +1,195 @@
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import org.databiosphere.workspacedata.model.RecordAttributes;
+import org.databiosphere.workspacedata.model.RecordRequest;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class RunCompletionHandler {
+  private final RunDao runDao;
+  private final WdsService wdsService;
+  private final ObjectMapper objectMapper;
+  private static final org.slf4j.Logger logger =
+      LoggerFactory.getLogger(RunCompletionHandler.class);
+
+  public RunCompletionHandler(RunDao runDao, WdsService wdsService, ObjectMapper objectMapper) {
+    this.runDao = runDao;
+    this.wdsService = wdsService;
+    this.objectMapper = objectMapper;
+  }
+
+  /*
+  The copy of SmartRunsPoller code.
+  Refactoring SmartRunsPoller will be covered in [WM-2090].
+   */
+  public boolean hasOutputDefinition(Run run) throws JsonProcessingException {
+    List<WorkflowOutputDefinition> outputDefinitionList =
+        objectMapper.readValue(run.runSet().outputDefinition(), new TypeReference<>() {});
+    return !outputDefinitionList.isEmpty();
+  }
+
+  /*
+  The copy of SmartRunsPoller code.
+  Refactoring SmartRunsPoller will be covered in [WM-2090].
+   */
+  public void updateOutputAttributes(Run run, Object outputs)
+      throws JsonProcessingException, WdsServiceException, CoercionException,
+          OutputProcessingException {
+
+    List<WorkflowOutputDefinition> outputDefinitionList =
+        objectMapper.readValue(run.runSet().outputDefinition(), new TypeReference<>() {});
+    RecordAttributes outputParamDef = OutputGenerator.buildOutputs(outputDefinitionList, outputs);
+    RecordRequest request = new RecordRequest().attributes(outputParamDef);
+
+    logger.info(
+        "Updating output attributes for Record ID {} from Run {}.", run.recordId(), run.engineId());
+
+    wdsService.updateRecord(request, run.runSet().recordType(), run.recordId());
+  }
+
+  public RunCompletionResult updateResults(
+      Run updatableRun, CbasRunStatus status, Object workflowOutputs, List<String> workflowErrors) {
+    long updateResultsStartNanos = System.nanoTime();
+    RunCompletionResult updateResult = RunCompletionResult.ERROR;
+
+    try {
+      var updatedRunState = status;
+
+      if (updatableRun.status() == updatedRunState && workflowOutputs == null) {
+        // Status is already up-to-date, there are no outputs to save.
+        // Only update last polled timestamp.
+        logger.info(
+            "Update last modified timestamp for Run {} (engine ID {}) in status {}.",
+            updatableRun.runId(),
+            updatableRun.engineId(),
+            updatableRun.status());
+
+        var changes = runDao.updateLastPolledTimestamp(updatableRun.runId());
+        if (changes != 1) {
+          logger.warn(
+              "Expected 1 row change updating last_polled_timestamp for Run {} in status {}, but got {}.",
```

Comment:
this warning looks like it was put here with certain expectations... could you help me understand better what situation this warning is capturing? (why do we expect exactly one row change, and why might there be more or less than one?)

Reply:
this is a check that the record has been actually updated. Since we do not have a full blown transactional support here (no Stairway), updating 0 records is possible situation (although very unlikely). This piece of code is moved/copied from SmartRunPoller.